From f6a123f06c28cd4c7ebe5ecf19c06b3d9118549d Mon Sep 17 00:00:00 2001
From: Benjamin Lin <benjamin-jw.lin@mediatek.com>
Date: Mon, 10 Mar 2025 16:56:53 +0800
Subject: [PATCH 06/92] mtk: mt76: mt7996: drop fragments with
 multicast/broadcast RA

IEEE 802.11 fragmentation can only be applied to unicast frames.
Hence, drop fragments with multicast/broadcast RA.
This patch fixes vulnerabilities, like CVE-2020-26145.

Signed-off-by: Benjamin Lin <benjamin-jw.lin@mediatek.com>
---
 mt7996/mac.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 771c16e8..0dbd4662 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -647,6 +647,14 @@ mt7996_mac_fill_rx(struct mt7996_dev *dev, enum mt76_rxq_id q,
 		status->last_amsdu = amsdu_info == MT_RXD4_LAST_AMSDU_FRAME;
 	}
 
+	/* IEEE 802.11 fragmentation can only be applied to unicast frames.
+	 * Hence, drop fragments with multicast/broadcast RA.
+	 * This check fixes vulnerabilities, like CVE-2020-26145.
+	 */
+	if ((ieee80211_has_morefrags(fc) || seq_ctrl & IEEE80211_SCTL_FRAG) &&
+	    FIELD_GET(MT_RXD3_NORMAL_ADDR_TYPE, rxd3) != MT_RXD3_NORMAL_U2M)
+		return -EINVAL;
+
 	hdr_gap = (u8 *)rxd - skb->data + 2 * remove_pad;
 	if (hdr_trans && ieee80211_has_morefrags(fc)) {
 		if (mt7996_reverse_frag0_hdr_trans(skb, hdr_gap))
-- 
2.45.2

